src/Compilers/CSharp/Portable/Symbols/Tuples/TupleFieldSymbol.cs

```diff
@@ -0,0 +1,233 @@
+            return result;
+        }
+
+        public override sealed int GetHashCode()
+        {
+            return Hash.Combine(_containingTuple.GetHashCode(), _tupleFieldId.GetHashCode());
+        }
+
+        public override sealed bool Equals(object obj)
+        {
+            return Equals(obj as TupleFieldSymbol);
+        }
+
+        public bool Equals(TupleFieldSymbol other)
+        {
+            if ((object)other == this)
+            {
+                return true;
+            }
+
+            return (object)other != null && _tupleFieldId == other._tupleFieldId && _containingTuple == other._containingTuple;
+        }
+    }
+
+    /// <summary>
+    /// Represents an element field of a tuple type (such as (int, byte).Item1)
+    /// that is backed by a real field with the same name within the tuple underlying type.
+    /// </summary>
+    internal class TupleElementFieldSymbol : TupleFieldSymbol
+    {
+        private readonly ImmutableArray<Location> _locations;
+
+        public TupleElementFieldSymbol(TupleTypeSymbol container, FieldSymbol underlyingField, int tupleFieldId, Location location)
+            : base(container, underlyingField, tupleFieldId)
+        {
+            _locations = location == null ? ImmutableArray<Location>.Empty : ImmutableArray.Create(location);
+        }
+
+        public override ImmutableArray<Location> Locations
+        {
+            get
+            {
+                return _locations;
+            }
+        }
+
+        public override ImmutableArray<SyntaxReference> DeclaringSyntaxReferences
+        {
+            get
+            {
+                return GetDeclaringSyntaxReferenceHelper<CSharpSyntaxNode>(_locations);
+            }
+        }
+
+        public override bool IsImplicitlyDeclared
+        {
+            get
+            {
+                return false;
+            }
+        }
+
+        internal override int? TypeLayoutOffset
+        {
+            get
+            {
+                if (_underlyingField.ContainingType != _containingTuple.TupleUnderlyingType)
+                {
+                    return null;
+                }
+
+                return base.TypeLayoutOffset;
+            }
+        }
+
+        public override Symbol AssociatedSymbol
+        {
+            get
+            {
+                if (_underlyingField.ContainingType != _containingTuple.TupleUnderlyingType)
+                {
+                    return null;
+                }
+
+                return base.AssociatedSymbol;
+            }
+        }
+    }
+
+    /// <summary>
+    /// Represents an element field of a tuple type (such as (int a, byte b).a, or (int a, byte b).b)
+    /// that is backed by a real field with a different name within the tuple underlying type.
+    /// </summary>
+    internal sealed class TupleRenamedElementFieldSymbol : TupleElementFieldSymbol
+    {
+        private readonly string _name;
```

Comment:
This type inherits equality from the base which does not factor in name.  Should this overridoe to factor in name?  Or is the index good enough?